File: algorithms/medium/optimal_partition_of_string.ex
defmodule Solution do
  @spec partition_string(s :: String.t()) :: integer
  def partition_string(s) do
    s
    |> String.graphemes()
    |> Enum.reduce({0, MapSet.new()}, fn c, {ct, set} ->
      if c in set do
        {ct + 1, MapSet.new([c])}
      else
        {ct, MapSet.put(set, c)}
      end
    end)
    |> elem(0)
    |> then(&(&1 + 1))
  end
end

File: algorithms/medium/optimal_partition_of_string.cpp
class Solution {
public:
  int partitionString(string s) {
    int ct = 0;
    unordered_set<char>set;
    for (auto& c : s) {
      if (set.find(c) != set.end()) {
        ct++;
        set.clear();
      }
      set.insert(c);
    }
    return ct + 1;
  }
};
